### ISSUE TITLE
Docs - Creating a new server

### ISSUE THREAD

**Issue by odscjames** [open-contracting/deploy#16]
https://ocdsdeploy.readthedocs.io/en/latest/creating-a-new-server.html

The current process depends on making entries in Open Data Services Coop resources and using the Open Data Services Coop deploy token. Document that fact here, or discuss a new process so that OCP staff can make new servers and document that here.

**Comment by odscjames** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Process & Docs - Creating a new server](https://trello.com/c/ZBQz7SDa/72-process-docs-creating-a-new-server)

**Comment by robredpath** [comment]
See also #11

**Comment by robredpath** [comment]
It's not actually explicit, but I assume it's implicit, that there's a task here to write up what's involved when you create a new server!

**Comment by odscjames** [comment]
* CRM pages could move to OCP CRM
* A keypass file could be created using OCP current practice.

We'll discuss and check we can move across to using these; then we can write up

**Comment by odscjames** [comment]
I have started this in https://ocdsdeploy.readthedocs.io/en/latest/how-to/create_server.html

Excuse committing straight to master as I was doing other work - also I was pretty confident I was putting it in the right place and we can always adjust content later as we do more.

There are some points we need to agree - these are in doc.

I put a section called Bytemark - other providers may be different. Maybe we'll want one section per provider or maybe we'll have some common sections to avoid repetition - but we can work that out later.

**Comment by odscjames** [comment]
Also, the section on Prometheus should  be expanded when we do  https://github.com/open-contracting/deploy/issues/31

**Comment by jpmckinney** [comment]
Looks good, I've added a few things. I've added a checklist in the issue description instead of leaving TODOs in the docs where we might forget about them.

Where is there documentation that python-msgpack and python-concurrent.futures are required? I can see old Salt issues referring to these, but I can't tell if they're still needed in the latest versions of salt / salt-ssh. Their [salt-ssh tutorial](https://docs.saltstack.com/en/getstarted/ssh/index.html) doesn't mention anything (not to say they aren't nonetheless needed).

**Comment by odscjames** [comment]
> Where is there documentation that python-msgpack and python-concurrent.futures are required?

I almost put a comment in the docs as I did this but wasn't sure that was the right place. I don't know; this is our old practice and it probably isn't needed any more. Both those packages are for Python 2 anyway.  I should have noted that here instead.

**Comment by odscjames** [comment]
> ODS CRM pages could move to OCP CRM wiki

I think all relevant content has now moved, so ticking.

> Logging the IP address, hostname, root password in a password manager (A keypass file could be created using OCP current practice)

I think this is the next thing to discuss to move this on - what's the best way to do that?

**Comment by jpmckinney** [comment]
Updating a KeePassX file has a lot of overhead (download the app, set up some way to sync the file with others, etc.).

Instead, I created a LastPass account, which OCP will only use for shared passwords between organizations. I'll send the user/pass via WhatsApp.

I added some previously shared credentials there.

**Comment by odscjames** [comment]
Thanks - I'll add anything we have and add a few bits to docs.

**Comment by odscjames** [comment]
Remaining ticks:

> Document how to generate and change root password

Think now done? 

> Add host-specific steps for Hetzner

I never actually set one up before, so I'm going to find that hard to write. Also, not sure if there cloud and real offerings are different. Do you want a placeholder with what should be achieved and some notes, and whoever next creates one can write up exact steps later?

> Logging the IP address, hostname, root password in a password manager (A keypass file could be created using OCP current practice)

We now have lastpass - are we concerned about making sure IP addresses are in there? 

> Check/Test whether python-msgpack and python-concurrent.futures are actually needed

Because different hosts might have different O.S. distributions, the only way I'd feel confident with this is spinning up a test Bytemark server for 30 mins. Is that ok?

**Comment by jpmckinney** [comment]
> Think now done?

Yes, now ticked.

> I never actually set one up before, so I'm going to find that hard to write. Also, not sure if there cloud and real offerings are different. Do you want a placeholder with what should be achieved and some notes, and whoever next creates one can write up exact steps later?

Who set up the Hetzner servers? We don't use their Cloud virtual servers, so can just write about their Robot real servers. Ideally the instructions will be as specific as their memory serves, and then otherwise it can be notes as you suggest.

> We now have lastpass - are we concerned about making sure IP addresses are in there?

I think `salt-config/roster` is enough for the IP address.

Taking a step back: What is the reason for storing the root password? Is it in case we misconfigure SSH, such that SSH key login no longer works?

> Because different hosts might have different O.S. distributions, the only way I'd feel confident with this is spinning up a test Bytemark server for 30 mins. Is that ok?

Yes, please go ahead.

Separately, I just noticed that an instruction is:

> Set Root user has to “SSH key (+ Password)” and enter your public key

What does Bytemark do with that public key? If it just adds it to `/root/.ssh/authorized_keys`, then we overwrite it (good). If it puts it elsewhere, then that user presumably has perpetual root access to the server (not so good).

**Comment by odscjames** [comment]
> Who set up the Hetzner servers?

@robredpath 

> What is the reason for storing the root password? Is it in case we misconfigure SSH, such that SSH key login no longer works?

Yes - Bytemark has console access through their web control panel

**Comment by odscjames** [comment]
>  If it just adds it to /root/.ssh/authorized_keys, 

Yes

> then we overwrite it (good).

Actually we use ssh_auth.present and ssh_auth.absent. They just edit the keys. So no.

>  then that user presumably has perpetual root access to the server (not so good).

Well ..... yes, but that person should be one of the sys admin people anyway.  I can't see a scenario where a person sets up a server who isn't a normal sysadmin root user. And if someone is doing that, they have the oppertunity to slip all kind of back doors on the server as they create it, so if we don't trust them we really shouldn't be letting them make new servers. So I don't see this as a problem at all.

**Comment by jpmckinney** [comment]
I'm thinking someone sets up the server, then leaves OCP, ODS, CDS, etc. They shouldn't retain access.

**Comment by odscjames** [comment]
> I'm thinking someone sets up the server, then at some later date leaves OCP, ODS, CDS, etc. They shouldn't retain access.

Agreed, but that's a different problem and applies to all root users, including the ones who don't set up servers. That's why we have root_authorized_keys_remove https://github.com/open-contracting/deploy/blob/master/salt/core.sls#L56 - anyone leaves (or anything else, a key is compromised) we can add it to root_to_remove file and remove from all servers quickly.

**Comment by odscjames** [comment]
> Check/Test whether python-msgpack and python-concurrent.futures are actually needed 

Just finished checking and it seems not on 18 LTS. I'll do a pull request.

**Comment by jpmckinney** [comment]
The point was just to check that Bytemark simply uses `/root/.ssh/authorized_keys`, which you say it does, which means the key is under our regular management, so we can handle it per usual.

**Comment by odscjames** [comment]
> The point was just to check that Bytemark simply uses /root/.ssh/authorized_keys for the "Set Root user has to “SSH key (+ Password)” and enter your public key" option, which you say it does, which means the key is under our regular management, so we can handle it per usual.

Ah, I see now. Yes, I checked it uses that file.

**Comment by jpmckinney** [comment]
Why do we change the root password set by Bytemark? Are their passwords less secure? Just wondering if we can skip that step.

**Comment by robredpath** [comment]
@jpmckinney AFAIK it's simply out of an abundance of caution; a password supplied by a third party could, in theory, have been stored, intercepted, be generated through an insecure algorithm, or something like that. For the sake of a small step in the process, we gain a small amount of extra security. 

More widely, I think that the practice helps to cultivate the 'distrust anything external' mindset that's useful - it's easy to forget about security when you're not thinking about security, but the mindset helps. 

All that said, if it's at all burdensome to use their password, it's probably fine!

**Comment by jpmckinney** [comment]
Makes sense! We'll keep the process as-is as it's just one more step.

**Comment by jpmckinney** [comment]
Closes by #161.
